pinot-core/src/main/java/com/linkedin/pinot/core/io/writer/impl/MmapMemoryManager.java

```diff
@@ -0,0 +1,121 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.linkedin.pinot.core.io.writer.impl;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.nio.channels.FileChannel;
+import java.util.LinkedList;
+import java.util.List;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import com.linkedin.pinot.common.segment.ReadMode;
+import com.linkedin.pinot.core.io.readerwriter.OffHeapMemoryManager;
+import com.linkedin.pinot.core.segment.memory.PinotDataBuffer;
+
+
+/**
+ * @class MmapMemoryManager is an OffHeapMemoryManager that manages memory by
+ * allocating it via mmap.
+ *
+ * This class attempts to minimize the overall number of file handles used for mapping the files.
+ * We create files of length 0.25g (or the requested buffer length, whichever is higher),
+ * and map areas of the file for each allocation request within a segment.
+ *
+ * @note Thread-unsafe. We expect to use this class only in a single writer case.
+ */
+public class MmapMemoryManager extends OffHeapMemoryManager {
+  private static final Logger LOGGER = LoggerFactory.getLogger(MmapMemoryManager.class);
+
+  private static final long FILE_LENGTH = 512 * 1024 * 1024L; // 0.5G per segment
+
+  private final String _dirPathName;
+  // It is possible that one thread stops consumption, and another thread takes over consumption, in LLC.
+  // So we should make numFiles as a volatile, in case buffer expansion happens in a different thread.
+  private volatile int _numFiles = 0;
+  private long _availableOffset = FILE_LENGTH; // Available offset in this file.
+  private long _curFileLen = -1;
+  private final List<String> _paths = new LinkedList<>();
+  PinotDataBuffer _currentBuffer;
+
+  /**
+   * @param dirPathName directory under which all mmap files are created.
+   * @param segmentName Name of the segment for which this memory manager allocates memory
+   */
+  public MmapMemoryManager(String dirPathName, String segmentName) {
+    super(segmentName);
+    _dirPathName = dirPathName;
+  }
+
+  private void addFileIfNecessary(long len) {
+    if (len + _availableOffset < _curFileLen) {
+      return;
+    }
+    String thisContext = getSegmentName() + "." + _numFiles++;
+    String filePath;
+    filePath = _dirPathName + "/" + thisContext;
+    final File file = new File(filePath);
+    if (file.exists()) {
+      throw new RuntimeException("File " + filePath + " already exists");
+    }
+    file.deleteOnExit();
+    RandomAccessFile raf;
+    try {
+      raf = new RandomAccessFile(filePath, "rw");
+    } catch (FileNotFoundException e) {
+      throw new RuntimeException(e);
+    }
+    long fileLen = Math.max(FILE_LENGTH, len);
```

Comment:
I want to pick a number that is high enough so that we can get multiple buffers, but low enoug that we don't run out of disk trying to ingest 100 partitions in a host. 2g will be an overkill, so I picked 0.5g. I think I should reduce it a bit, maybe.